tests/distributed/_test_distributed.py

```diff
@@ -0,0 +1,198 @@
+    }
+
+    def __init__(self, executable: str):
+        self.executable = executable
+
+    def worker_train(self, i: int) -> subprocess.CompletedProcess:
+        """Start the training process on the `i`-th worker."""
+        config_path = TESTS_DIR / f'train{i}.conf'
+        cmd = [self.executable, f'config={config_path}']
+        return subprocess.run(cmd)
+
+    def _set_ports(self) -> None:
+        """Randomly assign a port for training to each worker and save all ports to mlist.txt."""
+        ports = set(_generate_n_ports(self.n_workers))
+        i = 0
+        max_tries = 100
+        while i < max_tries and len(ports) < self.n_workers:
+            n_ports_left = self.n_workers - len(ports)
+            candidates = _generate_n_ports(n_ports_left)
+            ports.update(candidates)
+            i += 1
+        if i == max_tries:
+            raise RuntimeError('Unable to find non-colliding ports.')
+        self.listen_ports = list(ports)
+        with open(TESTS_DIR / 'mlist.txt', 'wt') as f:
+            for port in self.listen_ports:
+                f.write(f'127.0.0.1 {port}\n')
+
+    def _write_data(self, partitions: List[np.ndarray]) -> None:
+        """Write all training data as train.txt and each training partition as train{i}.txt."""
+        all_data = np.vstack(partitions)
+        np.savetxt(TESTS_DIR / 'train.txt', all_data, delimiter=',')
+        for i, partition in enumerate(partitions):
+            np.savetxt(TESTS_DIR / f'train{i}.txt', partition, delimiter=',')
+
+    def fit(self, partitions: List[np.ndarray], train_config: Dict = {}) -> None:
+        """Run the distributed training process on a single machine.
+
+        For each worker i:
+            1. The i-th partition is saved as train{i}.txt.
+            2. A random port is assigned for training.
+            3. A configuration file train{i}.conf is created.
+            4. The lightgbm binary is called with config=train{i}.conf in another thread.
+            5. The trained model is saved as model{i}.txt. Each model file only differs in data and local_listen_port.
+        The whole training set is saved as train.txt.
+        """
+        self.train_config = copy.deepcopy(self.default_train_config)
+        self.train_config.update(train_config)
+        self.n_workers = self.train_config['num_machines']
+        self._set_ports()
+        self._write_data(partitions)
+        self.label_ = np.hstack([partition[:, 0] for partition in partitions])
+        futures = []
+        with ThreadPoolExecutor(max_workers=self.n_workers) as executor:
+            for i in range(self.n_workers):
+                self.write_train_config(i)
+                train_future = executor.submit(self.worker_train, i)
+                futures.append(train_future)
+            results = [f.result() for f in futures]
+        for result in results:
+            if result.returncode != 0:
+                raise RuntimeError
```

Comment:
can you please add a specific error message here? Otherwise, if this error is triggered during testing we'll just see `pytest`'s logs end with an empty `RuntimeError` and will have to work backwards from the line number in the stack trace to figure out what happened.

Reply:
Since stderr from each worker is captured, pytest logs look like this:
![image](https://user-images.githubusercontent.com/8473587/123492486-3e06fa00-d5df-11eb-8eff-3d339630253d.png)
Should I add something like `'Error in training'`?

Reply:
yes please

Reply:
```suggestion
                raise RuntimeError('Error in training')
```

Since this was the only remaining suggestion on this PR, since it is so small, and since I'm agreeing with your suggestion, I feel comfortable using maintainer permissions to commit this change directly. I'll approve and merge once this builds one more time.